docs/development/awk/intro-to-awk/intro-to-awk.md

```diff
@@ -0,0 +1,732 @@
+
+### Looping in AWK
+
+AWK supports `for`, `do-while` and `while` loops that pretty much work like they
+do in the C programming language.
+
+#### An example
+
+The `ARGC` variable can help you iterate over the command line arguments of an AWK command
+with the help of a loop. This is illustrated in `loops.awk`, which is as follows:
+
+{{< file "loops.awk" awk >}}
+BEGIN {
+    for (i = 0; i < ARGC; i++)
+        printf "ARGV[%d] = %s\n", i, ARGV[i]
+
+	k = 0
+	while ( k < ARGC ) {
+        printf "ARGV[%d] = %s\n", k, ARGV[k]
+		k++
+	}
+
+	m = 0
+	do {
+        printf "ARGV[%d] = %s\n", m, ARGV[m]
+		m++
+	} while ( m < ARGC )
+
+}
+
+END {
+	for (i = 0; i < 10; i++)
+		printf "%d ", i
+	printf "\n"
+}
+{{< /file >}}
+
+Executing `loops.awk` will generate the following kind of output:
+
+    echo "" | awk -f loops.awk
+{{< output >}}
+ARGV[0] = awk
+ARGV[0] = awk
+ARGV[0] = awk
+0 1 2 3 4 5 6 7 8 9
+{{< /output >}}
+
+### Functions in AWK
+
+#### Predefined AWK functions
+
+AWK has built-in functions for string, number, time and date manipulation. You can
+easily generate random numbers using the `rand()` function. However `rand()` can
+generate repeatable output.
+
+The use of the `rand()` function is illustrated in the following code, which is saved in
+`rand.awk`:
+
+{{< file "rand.awk" awk >}}
+BEGIN {
+for (i=0; i<=10; i++) {
+	rnd[i] = 0;
+}
+
+while(i < 500)
+{
+	n = int(rand()*10);
+	rnd[n]++;
+	i++;
+}
+
+for( i=0; i<10; i++) {
+	print i, "Occurred", rnd[i], "times";
+}
+}
+{{< /file >}}
+
+If you recall from the `hw.awk` program, the use of `BEGIN` is mandatory in order
+to avoid waiting for input, which is not required in this program. Apart from this,
+you just have two `for` loops and a single `while` loop.
+
+Executing `rand.awk` will generate the following kind of output:
+
+    awk -f rand.awk
+{{< output >}}
+0 Occurred 42 times
+1 Occurred 50 times
+2 Occurred 52 times
+3 Occurred 51 times
+4 Occurred 43 times
+5 Occurred 44 times
+6 Occurred 64 times
+7 Occurred 47 times
+8 Occurred 49 times
+9 Occurred 47 times
+{{< /output >}}
+
+### User Defined functions in AWK 
+
+The AWK programming language allows you to define your own functions and use them
+in your AWK programs. This is illustrated in `myFunction.awk`, which is as follows:
+
+{{< file "myFunction.awk" awk >}}
+function isnum(x) { return(x==x+0) }
+
+function sumToN(n)
+{
+    sum = 0
+    if (n < 0) { n = -n }
+    if ( isnum(n) )
+    {
+        for (j = 1; j <= n; j++) { sum = sum + j }
+    }
+    else { return -1 }
+    return sum
+}
+
+{
+    for (i=1; i<=NF; i++)
+    {
+        print $i, "\t:", sumToN($i)
+    }
+}
+{{< /file >}}
+
+The user defined function is called `sumToN()`. The `isnum()` function is for making
+sure that the input to `sumToN()` will be numeric.
+
+Executing `myFunction.awk` will generate the following kind of output:
+
+    echo "10 12" | awk -f myFunction.awk
+{{< output >}}
+10 	: 55
+12 	: 78
+{{< /output >}}
+
+### BEGIN and END
```

Comment:
See notes above about restructuring the explanation of patterns in AWK

Reply:
I have added an explanation about patters earlier on.